### PR TITLE
fix: unscale fp16 gradient problem & potential error (#6086)

### PR DESCRIPTION
**Pull request by lvzii**
Referring to [commit](https://github.com/lvzii/diffusers/commit/288ceebea51abb439ca81684d642852c7fcc044a) ,
Fixes # (6086) in `train_text_to_image_lora_sdxl.py`, and Fixes # (4619) in `train_text_to_image_lora_sdxl.py` during fixing the former error.

### CODE REVIEW

File: examples/text_to_image/train_text_to_image_lora_sdxl.py [hunk 1]
```diff
@@ -1187,6 +1198,9 @@ def compute_time_ids(original_size, crops_coords_top_left):
         torch.cuda.empty_cache()
 
         # Final inference
+        # Make sure vae.dtype is consistent with the unet.dtype
+        if args.mixed_precision == "fp16":
+            vae.to(weight_dtype)
```
Comment (sayakpaul) [inline examples/text_to_image/train_text_to_image_lora_sdxl.py:1203]: This is not needed in my opinion. We already set the `torch_dtype` in the pipeline during loading it.

File: examples/text_to_image/train_text_to_image_lora_sdxl.py [hunk 2]
```diff
@@ -640,6 +640,17 @@ def main(args):
...
+    # Make sure the trainable params are in float32.
+    if args.mixed_precision == "fp16":
+        models = [unet]
+        if args.train_text_encoder:
+            models.extend([text_encoder_one, text_encoder_two])
+        for model in models:
+            for param in model.parameters():
+                # only upcast trainable parameters (LoRA) into fp32
+                if param.requires_grad:
+                    param.data = param.to(torch.float32)
```
Comment (sayakpaul) [inline examples/text_to_image/train_text_to_image_lora_sdxl.py:652]: Works for me!

**Review by sayakpaul — APPROVED** [review]
TYSM!

File: examples/text_to_image/train_text_to_image_lora_sdxl.py (hunk 1 above)
Comment (lvzii) [reply examples/text_to_image/train_text_to_image_lora_sdxl.py:1203]: If no other `pretrained_vae_model_name_or_path` is set, then the vae is set to float32,
```
if args.pretrained_vae_model_name_or_path is None:
        vae.to(accelerator.device, dtype=torch.float32)
```
and the pipeline here does not reload vae. So `vae.dtype` `float32` != `unet.dtype` `fp16`, which in my tests causes `RuntimeError: Input type (c10::Half) and bias type (float) should be the same`

Comment (sayakpaul) [reply examples/text_to_image/train_text_to_image_lora_sdxl.py:1203]: Ah okay got it!

**Comment by sayakpaul** [comment]
Thank you for your contributions!

**Pull request merged**